File: eggs/Nougat/src/main/java/com/android_n/egg/preview/PlatLogoActivity.java
package com.android_n.egg.preview;

import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.View;
import android.view.animation.PathInterpolator;
import android.widget.FrameLayout;
import android.widget.ImageView;

public class PlatLogoActivity extends Activity {

    FrameLayout mLayout;
    int mTapCount;
    int mKeyCount;

    PathInterpolator mInterpolator = new PathInterpolator(0f, 0f, 0.5f, 1f);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLayout = new FrameLayout(this);
        setContentView(mLayout);
    }

    @Override
    public void onAttachedToWindow() {
        final DisplayMetrics dm = getResources().getDisplayMetrics();
        final float dp = dm.density;

        final int size = (int)
                (Math.min(Math.min(dm.widthPixels, dm.heightPixels), 600 * dp) - 100 * dp);

        final ImageView im = new ImageView(this);

        final int pad = (int) (40 * dp);
        im.setPadding(pad, pad, pad, pad);
        im.setTranslationZ(20);
        im.setScaleX(0.5f);
        im.setScaleY(0.5f);
        im.setAlpha(0f);

        Drawable logo = getDrawable(R.drawable.n_android_logo);

        im.setBackground(new RippleDrawable(
                ColorStateList.valueOf(0xFFFFFFFF),
                logo,
                null
        ));

        im.setClickable(true);

        im.setOnClickListener(v -> {
            mTapCount++;

            im.setOnLongClickListener(v1 -> {
                if (mTapCount < 5) return false;

                Log.d("PlatLogo", "Easter egg triggered!");

                try {
                    startActivity(new Intent(Intent.ACTION_MAIN));
                } catch (Exception e) {
                    Log.e("PlatLogo", "No activity", e);
                }

                return true;
            });
        });

        // Keyboard support
        im.setFocusable(true);
        im.requestFocus();

        im.setOnKeyListener((v, keyCode, event) -> {
            if (keyCode != KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN) {
                mKeyCount++;
                if (mKeyCount > 2) {
                    if (mTapCount > 5) {
                        im.performLongClick();
                    } else {
                        im.performClick();
                    }
                }
                return true;
            }
            return false;
        });

        mLayout.addView(im, new FrameLayout.LayoutParams(size, size, Gravity.CENTER));

        im.animate()
                .scaleX(1f)
                .scaleY(1f)
                .alpha(1f)
                .setInterpolator(mInterpolator)
                .setDuration(500)
                .setStartDelay(800)
                .start();
    }
}                            

File: eggs/KitKat/src/main/java/com/android_k/egg/PlatLogoActivity.kt
package com.android_k.egg

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.AnticipateOvershootInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import com.android_k.egg.SpUtils
import com.dede.basic.utils.TransformationMethodUtils

class PlatLogoActivity : Activity() {

    private lateinit var mContent: FrameLayout

    companion object {
        const val BGCOLOR = 0xffed1d24.toInt()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)

        val bold = Typeface.create("sans-serif", Typeface.BOLD)
        val light = Typeface.create("sans-serif-light", Typeface.NORMAL)

        mContent = FrameLayout(this).apply {
            setBackgroundColor(0xC0000000.toInt())
        }

        val lp = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        ).apply {
            gravity = Gravity.CENTER
        }

        val logo = ImageView(this).apply {
            setImageResource(R.drawable.k_platlogo)
            scaleType = ImageView.ScaleType.CENTER_INSIDE
            visibility = View.INVISIBLE
        }

        val bg = View(this).apply {
            setBackgroundColor(BGCOLOR)
            alpha = 0f
        }

        val letter = TextView(this).apply {
            typeface = bold
            textSize = 300f
            setTextColor(0xFFFFFFFF.toInt())
            gravity = Gravity.CENTER
            text = "K"
        }

        val p = (4 * metrics.density).toInt()

        val tv = TextView(this).apply {
            light?.let { typeface = it }
            textSize = 30f
            setPadding(p, p, p, p)
            setTextColor(0xFFFFFFFF.toInt())
            gravity = Gravity.CENTER
            transformationMethod =
                TransformationMethodUtils.createAllCapsTransformationMethod(this@PlatLogoActivity)
            text = "Android 4.4"
            visibility = View.INVISIBLE
        }

        mContent.apply {
            addView(bg)
            addView(letter, lp)
            addView(logo, lp)

            val lp2 = FrameLayout.LayoutParams(lp).apply {
                gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
                bottomMargin = (10f * p).toInt()
            }

            addView(tv, lp2)
        }

        var clicks = 0

        mContent.setOnClickListener {
            clicks++
            if (clicks >= 6) {
                mContent.performLongClick()
                return@setOnClickListener
            }

            letter.animate().cancel()
            val offset = letter.rotation.toInt() % 360

            letter.animate()
                .rotationBy((if (Math.random() > 0.5f) 360f else -360f) - offset)
                .setInterpolator(DecelerateInterpolator())
                .setDuration(700)
                .start()
        }

        mContent.setOnLongClickListener {
            if (logo.visibility != View.VISIBLE) {

                bg.scaleX = 0.01f
                bg.animate().alpha(1f).scaleX(1f).setStartDelay(500).start()

                letter.animate()
                    .alpha(0f)
                    .scaleY(0.5f)
                    .scaleX(0.5f)
                    .rotationBy(360f)
                    .setInterpolator(AccelerateInterpolator())
                    .setDuration(1000)
                    .start()

                logo.apply {
                    alpha = 0f
                    visibility = View.VISIBLE
                    scaleX = 0.5f
                    scaleY = 0.5f
                    animate()
                        .alpha(1f)
                        .scaleX(1f)
                        .scaleY(1f)
                        .setDuration(1000)
                        .setStartDelay(500)
                        .setInterpolator(AnticipateOvershootInterpolator())
                        .start()
                }

                tv.apply {
                    alpha = 0f
                    visibility = View.VISIBLE
                    animate()
                        .alpha(1f)
                        .setDuration(1000)
                        .setStartDelay(1000)
                        .start()
                }

                true
            } else false
        }

        logo.setOnLongClickListener {
            if (SpUtils.getLong(it.context, "k_egg_mode", 0) == 0L) {
                SpUtils.putLong(
                    it.context,
                    "k_egg_mode",
                    System.currentTimeMillis()
                )
            }

            try {
                startActivity(Intent(this, DessertCase::class.java))
            } catch (ex: ActivityNotFoundException) {
                android.util.Log.e("PlatLogoActivity", "Couldn't catch a break.")
            }

            finish()
            true
        }

        setContentView(mContent)
    }
}
